File: chroma-league/src/main/java/com/bonepl/chromaleague/state/EventData.java
package com.bonepl.chromaleague.state;

import com.bonepl.chromaleague.rest.eventdata.DragonType;
import com.bonepl.chromaleague.rest.eventdata.Event;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

public class EventData {
    private final List<DragonType> killedDragons = new ArrayList<>();
    private final List<Event> processedEvents = new LinkedList<>();
    private Double baronBuffEnd;
    private Double elderBuffEnd;
    private int totalEldersKilled;
    private int activePlayerKillingSpree;
    private int activePlayerAssistSpree;
    private boolean riftAnimationPlayed = false;
    private RespawnIndicator respawnIndicator = RespawnIndicator.IDLE;
    private Double deathTime;
    private Double respawnTime;

    public Double getBaronBuffEnd() {
        return baronBuffEnd;
    }

    public void setBaronBuffEnd(Double baronBuffEnd) {
        this.baronBuffEnd = baronBuffEnd;
    }

    public List<DragonType> getKilledDragons() {
        return Collections.unmodifiableList(killedDragons);
    }

    public void addKilledDragon(DragonType dragonType) {
        killedDragons.add(dragonType);
    }

    public Double getElderBuffEnd() {
        return elderBuffEnd;
    }

    public void setElderBuffEnd(Double elderBuffEnd) {
        this.elderBuffEnd = elderBuffEnd;
    }

    public int getTotalEldersKilled() {
        return totalEldersKilled;
    }

    public void setTotalEldersKilled(int totalEldersKilled) {
        this.totalEldersKilled = totalEldersKilled;
    }

    public int getActivePlayerKillingSpree() {
        return activePlayerKillingSpree;
    }

    public void setActivePlayerKillingSpree(int activePlayerKillingSpree) {
        this.activePlayerKillingSpree = activePlayerKillingSpree;
    }

    public int getActivePlayerAssistSpree() {
        return activePlayerAssistSpree;
    }

    public void setActivePlayerAssistSpree(int activePlayerAssistSpree) {
        this.activePlayerAssistSpree = activePlayerAssistSpree;
    }

    public boolean didRiftAnimationPlay() {
        return riftAnimationPlayed;
    }

    public void setRiftAnimationPlayed(boolean riftAnimationPlayed) {
        this.riftAnimationPlayed = riftAnimationPlayed;
    }

    public void addProcessedEvents(List<Event> events) {
        processedEvents.addAll(events);
    }

    public List<Event> getUnprocessedEvents(List<Event> fetchedEvents) {
        return fetchedEvents.stream()
                .filter(fetchedEvent -> !processedEvents.contains(fetchedEvent)).toList();
    }

    public RespawnIndicator getRespawnIndicator() {
        return respawnIndicator;
    }

    public void setRespawnIndicator(RespawnIndicator respawnIndicator) {
        this.respawnIndicator = respawnIndicator;
    }

    public Double getDeathTime() {
        return deathTime;
    }

    public void setDeathTime(Double deathTime) {
        this.deathTime = deathTime;
    }

    public Double getRespawnTime() {
        return respawnTime;
    }

    public void setRespawnTime(Double respawnTime) {
        this.respawnTime = respawnTime;
    }

    //TEST ONLY
    public List<Event> getProcessedEvents() {
        return Collections.unmodifiableList(processedEvents);
    }
}

File: src/state/EventData.ts
import {DragonType} from '../rest/eventdata/DragonType.js';
import {Event} from '../rest/eventdata/Event.js';
import {RespawnIndicator} from './RespawnIndicator.js';

export class EventData {
    private readonly killedDragons: DragonType[] = [];
    private readonly processedEvents: Event[] = [];
    private readonly processedEventIds = new Set<number>();
    private lastProcessedEventTime: number | null = null;
    baronBuffEnd: number | null = null;
    elderBuffEnd: number | null = null;
    totalEldersKilled = 0;
    activePlayerKillingSpree = 0;
    activePlayerAssistSpree = 0;
    riftAnimationPlayed = false;
    respawnIndicator: RespawnIndicator = RespawnIndicator.IDLE;
    deathTime: number | null = null;
    respawnTime: number | null = null;

    getKilledDragons(): readonly DragonType[] {
        return this.killedDragons;
    }

    addKilledDragon(dragonType: DragonType): void {
        this.killedDragons.push(dragonType);
    }

    getLastProcessedEventTime(): number | null {
        return this.lastProcessedEventTime;
    }

    addProcessedEvents(events: Event[]): void {
        for (const event of events) {
            this.processedEvents.push(event);
            this.processedEventIds.add(event.EventID);
            if (this.lastProcessedEventTime === null || event.EventTime > this.lastProcessedEventTime) {
                this.lastProcessedEventTime = event.EventTime;
            }
        }
    }

    getUnprocessedEvents(fetchedEvents: Event[]): Event[] {
        return fetchedEvents.filter(fetched => !this.processedEventIds.has(fetched.EventID));
    }

    /** Test only */
    getProcessedEvents(): readonly Event[] {
        return this.processedEvents;
    }
}
